### PR TITLE
add support for additional arbitrary tweaks to the verifying key, par…

### PR DESCRIPTION
**Pull request by arminsabouri**
…tial signatures and signature aggregation

### CODE REVIEW

File: frost-core/src/lib.rs [hunk 1]
```diff
@@ -365,6 +365,9 @@ pub struct SigningPackage<C: Ciphersuite> {
         )
     )]
     message: Vec<u8>,
+    /// Additional tweak for both the secret share and the verifying key
+    #[cfg_attr(feature = "serde", serde(skip))]
```
Comment (stevenroose) [inline frost-core/src/lib.rs]: Shouldn't skip the serialization of this

File: frost-core/src/round2.rs [hunk 2]
```diff
@@ -92,17 +92,19 @@ where
         challenge: &Challenge<C>,
         group_commitment: &frost::GroupCommitment<C>,
         verifying_key: &frost::VerifyingKey<C>,
+        additional_tweak: &Option<Vec<u8>>,
```
Comment (stevenroose) [inline frost-core/src/round2.rs]: `Option<&[u8]>` is kinda identical but more idiomatic

File: frost-core/src/signing_key.rs [hunk 3]
```diff
@@ -44,11 +46,16 @@ where
     }
 
     /// Create a signature `msg` using this `SigningKey`.
-    pub fn sign<R: RngCore + CryptoRng>(&self, mut rng: R, msg: &[u8]) -> Signature<C> {
+    pub fn sign<R: RngCore + CryptoRng>(
+        &self,
+        mut rng: R,
+        msg: &[u8],
+        additional_tweak: &Option<Vec<u8>>,
```
Comment (stevenroose) [inline frost-core/src/signing_key.rs]: `Option<&[u8]>` (btw you can get it at call-site by using `.as_ref()`

File: frost-core/src/traits.rs [hunk 4]
```diff
@@ -267,8 +272,8 @@ pub trait Ciphersuite: Copy + Clone + PartialEq + Debug {
         z: <<Self::Group as Group>::Field as Field>::Scalar,
         challenge: &Challenge<Self>,
         verifying_key: &Element<Self>,
-    ) -> <<Self::Group as Group>::Field as Field>::Scalar
-    {
+        arbitrary_tweak: &Option<Vec<u8>>,
```
Comment (stevenroose) [inline frost-core/src/traits.rs]: `Option<&[u8]>`

File: frost-core/src/traits.rs [hunk 5]
```diff
@@ -279,8 +284,8 @@ pub trait Ciphersuite: Copy + Clone + PartialEq + Debug {
         secret: <<Self::Group as Group>::Field as Field>::Scalar,
         challenge: <<Self::Group as Group>::Field as Field>::Scalar,
         verifying_key: &Element<Self>,
-    ) -> <<Self::Group as Group>::Field as Field>::Scalar
-    {
+        arbitrary_tweak: &Option<Vec<u8>>,
```
Comment (stevenroose) [inline frost-core/src/traits.rs]: `Option<&[u8]>`

File: frost-core/src/traits.rs [hunk 6]
```diff
@@ -293,15 +298,16 @@ pub trait Ciphersuite: Copy + Clone + PartialEq + Debug {
         lambda_i: <<Self::Group as Group>::Field as Field>::Scalar,
         key_package: &crate::keys::KeyPackage<Self>,
         challenge: Challenge<Self>,
-    ) -> crate::round2::SignatureShare<Self>
-    {
+        arbitrary_tweak: &Option<Vec<u8>>,
```
Comment (stevenroose) [inline frost-core/src/traits.rs]: `Option<&[u8]>`

File: frost-core/src/traits.rs [hunk 7]
```diff
@@ -293,15 +298,16 @@ pub trait Ciphersuite: Copy + Clone + PartialEq + Debug {
...
+    ) -> crate::round2::SignatureShare<Self> {
         panic!("Not implemented");
     }
 
     /// calculate tweaked public key (used in frost-sepc256k1-tr)
     #[allow(unused)]
     fn tweaked_public_key(
         public_key: &<Self::Group as Group>::Element,
+        arbitrary_tweak: &Option<Vec<u8>>,
```
Comment (stevenroose) [inline frost-core/src/traits.rs]: `Option<&[u8]>`

File: frost-core/src/traits.rs [hunk 8]
```diff
@@ -319,8 +325,8 @@ pub trait Ciphersuite: Copy + Clone + PartialEq + Debug {
     fn tweaked_secret_key(
         secret: <<Self::Group as Group>::Field as Field>::Scalar,
         public: &Element<Self>,
-    ) -> <<Self::Group as Group>::Field as Field>::Scalar
-    {
+        additional_tweak: &Option<Vec<u8>>,
```
Comment (stevenroose) [inline frost-core/src/traits.rs]: `Option<&[u8]>`

File: frost-core/src/traits.rs [hunk 9]
```diff
@@ -349,8 +353,8 @@ pub trait Ciphersuite: Copy + Clone + PartialEq + Debug {
     fn taproot_compat_verifying_share(
         verifying_share: &<Self::Group as Group>::Element,
         verifying_key: &<Self::Group as Group>::Element,
-    ) -> <Self::Group as Group>::Element
-    {
+        additional_tweak: &Option<Vec<u8>>,
```
Comment (stevenroose) [inline frost-core/src/traits.rs]: `Option<&[u8]>`

File: frost-core/src/verifying_key.rs [hunk 10]
```diff
@@ -63,6 +63,7 @@ where
         &self,
         challenge: Challenge<C>,
         signature: &Signature<C>,
+        additional_tweak: &Option<Vec<u8>>,
```
Comment (stevenroose) [inline frost-core/src/verifying_key.rs]: `Option<&[u8]>`

File: frost-core/src/verifying_key.rs [hunk 11]
```diff
@@ -86,8 +87,13 @@ where
     }
 
     /// Verify a purported `signature` over `msg` made by this verification key.
-    pub fn verify(&self, msg: &[u8], signature: &Signature<C>) -> Result<(), Error<C>> {
-        C::verify_signature(msg, signature, self)
+    pub fn verify(
+        &self,
+        msg: &[u8],
+        signature: &Signature<C>,
+        additional_tweak: &Option<Vec<u8>>,
```
Comment (stevenroose) [inline frost-core/src/verifying_key.rs]: `Option<&[u8]>`

File: frost-secp256k1-tr/src/lib.rs [hunk 12]
```diff
@@ -194,6 +194,17 @@ pub fn hasher_to_scalar(hasher: Sha256) -> Scalar {
     Scalar::from(&sp)
 }
 
+/// Take arbitrary data and convert it to a scalar
+pub fn additional_tweak_scalar(
+    public_key: &<<Secp256K1Sha256 as Ciphersuite>::Group as Group>::Element,
+    data: &Vec<u8>,
```
Comment (stevenroose) [inline frost-secp256k1-tr/src/lib.rs]: `&[u8]`

File: frost-secp256k1-tr/src/lib.rs [hunk 13]
```diff
@@ -220,20 +231,29 @@ pub fn tweak(
 pub fn tweaked_public_key(
     public_key: &<<Secp256K1Sha256 as Ciphersuite>::Group as Group>::Element,
     merkle_root: &[u8],
+    additional_tweak: &Option<Vec<u8>>,
```
Comment (stevenroose) [inline frost-secp256k1-tr/src/lib.rs]: `Option<&[u8]>`

File: frost-secp256k1-tr/src/lib.rs [hunk 14]
```diff
@@ -220,20 +231,29 @@ pub fn tweak(
...
+    }
+
     ProjectivePoint::GENERATOR * tweak(&pk, merkle_root) + pk
 }
 
 /// Creates a real BIP341 tweaked public key by assuming an even y-coordinate.
 pub fn real_tweaked_pubkey(
     public_key: &<<Secp256K1Sha256 as Ciphersuite>::Group as Group>::Element,
     merkle_root: &[u8],
+    additional_tweak: &Option<Vec<u8>>,
```
Comment (stevenroose) [inline frost-secp256k1-tr/src/lib.rs]: `Option<&[u8]>`

File: frost-secp256k1-tr/src/lib.rs [hunk 15]
```diff
@@ -244,11 +264,20 @@ pub fn tweaked_secret_key(
     secret: <<<Secp256K1Sha256 as Ciphersuite>::Group as Group>::Field as Field>::Scalar,
     public_key: &<<Secp256K1Sha256 as Ciphersuite>::Group as Group>::Element,
     merkle_root: &[u8],
+    additional_tweak: &Option<Vec<u8>>,
```
Comment (stevenroose) [inline frost-secp256k1-tr/src/lib.rs]: `Option<&[u8]>`

File: frost-secp256k1-tr/src/lib.rs [hunk 16]
```diff
@@ -315,9 +344,15 @@ impl Ciphersuite for Secp256K1Sha256 {
     }
 
     /// Generates the challenge as is required for Schnorr signatures.
-    fn challenge(R: &Element<S>, verifying_key: &VerifyingKey, msg: &[u8]) -> Challenge<S> {
+    fn challenge(
+        R: &Element<S>,
+        verifying_key: &VerifyingKey,
+        msg: &[u8],
+        additional_tweak: &Option<Vec<u8>>,
```
Comment (stevenroose) [inline frost-secp256k1-tr/src/lib.rs]: `Option<&[u8]>`

File: frost-secp256k1-tr/src/lib.rs [hunk 17]
```diff
@@ -334,10 +369,16 @@ impl Ciphersuite for Secp256K1Sha256 {
         z: <<Self::Group as Group>::Field as Field>::Scalar,
         challenge: &Challenge<S>,
         verifying_key: &Element<S>,
+        additional_tweak: &Option<Vec<u8>>,
```
Comment (stevenroose) [inline frost-secp256k1-tr/src/lib.rs]: `Option<&[u8]>`

File: frost-secp256k1-tr/src/lib.rs [hunk 18]
```diff
@@ -351,8 +392,9 @@ impl Ciphersuite for Secp256K1Sha256 {
         secret: <<Self::Group as Group>::Field as Field>::Scalar,
         challenge: <<Self::Group as Group>::Field as Field>::Scalar,
         verifying_key: &Element<S>,
+        additional_tweak: &Option<Vec<u8>>,
```
Comment (stevenroose) [inline frost-secp256k1-tr/src/lib.rs]: `Option<&[u8]>`

File: frost-secp256k1-tr/src/lib.rs [hunk 19]
```diff
@@ -368,6 +410,7 @@ impl Ciphersuite for Secp256K1Sha256 {
         lambda_i: <<Self::Group as Group>::Field as Field>::Scalar,
         key_package: &frost::keys::KeyPackage<S>,
         challenge: Challenge<S>,
+        additional_tweak: &Option<Vec<u8>>,
```
Comment (stevenroose) [inline frost-secp256k1-tr/src/lib.rs]: `Option<&[u8]>`

File: frost-secp256k1-tr/src/lib.rs [hunk 20]
```diff
@@ -391,8 +435,9 @@ impl Ciphersuite for Secp256K1Sha256 {
     /// calculate tweaked public key
     fn tweaked_public_key(
         public_key: &<Self::Group as Group>::Element,
+        additional_tweak: &Option<Vec<u8>>,
```
Comment (stevenroose) [inline frost-secp256k1-tr/src/lib.rs]: `Option<&[u8]>`

File: frost-secp256k1-tr/src/lib.rs [hunk 21]
```diff
@@ -436,10 +482,11 @@ impl Ciphersuite for Secp256K1Sha256 {
     fn taproot_compat_verifying_share(
         verifying_share: &<Self::Group as Group>::Element,
         verifying_key: &<Self::Group as Group>::Element,
+        arbitrary_tweak: &Option<Vec<u8>>,
```
Comment (stevenroose) [inline frost-secp256k1-tr/src/lib.rs]: `Option<&[u8]>`

**Review by stevenroose — CHANGES_REQUESTED** [review]
sorry, ended up repeating that comment a ton of times :/

**Comment by arminsabouri** [comment]
> sorry, ended up repeating that comment a ton of times :/

@stevenroose all good. I actually wanted specific feedback on that.

**Comment by arminsabouri** [comment]
Closing replaced by #2

**Pull request closed without merging**